### PR TITLE
Make `runParser()` and `runWith*()` parser-aware for help, version, and completion

### LINKED ISSUE #230: The `runParser()` and `runWith*()` runners treat meta option names as runner requests even when they are ordinary parser values
## Summary

Runner-level help/version/completion detection currently scans raw argument tokens without respecting parser value boundaries, so tokens like `--help`, `--version`, and `--completion` can be hijacked as runner meta requests even when they are simply ordinary parser values.

## Documentation context

`docs/concepts/runners.md` presents `parse()` and `runParser()` as adjacent layers over the same parser definitions, and `docs/concepts/primitives.md` says arguments after `--` or option values should be treated as ordinary parser data. That makes it reasonable to expect runner meta handling to preserve parser value boundaries rather than reinterpreting data tokens as out-of-band control requests.

## Reproduction

A minimal option-value repro is `const parser = object({ message: option("--message", string({ metavar: "MESSAGE" })) });`, where plain parser semantics accept all of `parse(parser, ["--message", "--help"])`, `parse(parser, ["--message", "--version"])`, and `parse(parser, ["--message", "--completion"])` successfully as ordinary string values.

## Additional cases

The same class of problem appears for ordinary positional values. For example, `const parser = object({ value: argument(string({ metavar: "VALUE" })) });` accepts `parse(parser, ["--completion=bash"])` successfully, but `runParser(parser, "myapp", ["--completion=bash"], { completion: { option: true, onShow: () => "COMPLETION" } })` and the equivalent `runWithSync(...)` call both go to completion instead of returning the parsed value.

Plain positional words that collide with command-form meta names are also affected. With `help`, `version`, or `completion` command support enabled, inputs such as `["help"]`, `["version"]`, `["completion"]`, or custom aliases like `["assist"]`, `["ver"]`, and `["completions"]` are reclassified as meta commands instead of being delivered to `argument(string())` as ordinary parser data.

Custom meta option names are affected too. If help/version/completion are configured with names like `-h`, `-V`, or `--completions`, then `runParser(parser, "myapp", ["--message", "-h"], ...)`, `runParser(parser, "myapp", ["--message", "-V"], ...)`, and `runParser(parser, "myapp", ["--message", "--completions"], ...)` reclassify those ordinary values as runner meta requests in the same way.

The same bug is visible through `@optique/run` wrappers. For example, `runSync(parser, { programName: "myapp", args: ["--message", "--help"], help: "option", onExit: () => "EXIT" })` shows help instead of preserving the ordinary option value.

## Actual behavior

Runner meta detection overrides ordinary parser values and treats them as out-of-band help/version/completion requests.

## Expected behavior

Once a token is syntactically being used as parser data, runner meta handling should not reinterpret it as a global request. The result should stay aligned with ordinary parser semantics.

## Notes

`packages/core/src/facade.test.ts` already establishes nearby control cases:

 - last-option-wins behavior for ordinary help/version conflicts is explicitly tested
 - `needsEarlyExit does not trigger for non-matching completion option args` shows the fast path is supposed to stay out of the way when the token stream is not a real meta request

That narrows this issue to missing parser-boundary awareness, not to meta handling in general.

The root cause is the raw token scan in `packages/core/src/facade.ts`.

 - `needsEarlyExit()` uses `args.includes(...)`-style checks and direct string scans for help/version/completion names with no awareness of option-value boundaries, positional parsing state, or `--` termination.
 - the lenient meta handling and later `classifyResult(...)` logic also reason from raw `args` positions such as `args[0]` and `args.includes(...)` instead of parser-consumed structure.

So the runner layer is effectively doing a second, weaker parse of the token stream for meta features, and that second parse can steal tokens that the user parser would otherwise treat as ordinary values.

### PR DESCRIPTION
**Pull request by dahlia**
Closes https://github.com/dahlia/optique/issues/230.

This PR fixes runner-level help, version, and completion detection so it follows parser boundaries instead of scanning raw `argv` and stealing tokens that are ordinary parser values. The main change is in *packages/core/src/facade.ts*, where `runParser()`, `runWith()`, and `runWithSync()` now parse with the user parser first and classify only genuinely unconsumed tokens as meta requests.

This changes visible behavior in the cases reported in the issue. Inputs such as `["--message", "--help"]`, `["help"]`, and `["--completion=bash"]` now stay aligned with ordinary parser semantics when those tokens are valid data for the user parser, while real meta requests such as top-level `--help`, `serve --help`, `--version`, and completion requests still behave as before.

```ts
const parser = object({
  message: option("--message", string({ metavar: "MESSAGE" })),
});

runParser(parser, "myapp", ["--message", "--help"], {
  help: { option: true },
});
// => { message: "--help" }
```

```ts
const parser = object({
  value: argument(string({ metavar: "VALUE" })),
});

runParser(parser, "myapp", ["help"], {
  help: { command: true },
});
// => { value: "help" }
```

To keep subcommand help and completion accurate after parser-first classification, this PR threads matched command names through the execution context in *packages/core/src/parser.ts*, *packages/core/src/primitives.ts*, *packages/core/src/constructs.ts*, and *packages/core/src/modifiers.ts*. I also removed the old startup validation rule that rejected user parsers for reusing built-in meta names and aliases, because that rule no longer matches the intended runtime semantics. A follow-up cleanup trims the obsolete validation plumbing in *packages/core/src/validate.ts* and its focused tests.

I updated *docs/concepts/runners.md*, *docs/concepts/extend.md*, *docs/integrations/config.md*, *docs/integrations/env.md*, *packages/run/README.md*, and *CHANGES.md* to document the new parser-aware runner semantics and the user-visible behavior change after 0.10.0.

Regression coverage was added in *packages/core/src/facade.test.ts*, *packages/core/src/validate.test.ts*, and *packages/run/src/run.test.ts*. I also reran the full project test matrix with `mise test` and rebuilt the docs with `pnpm -C docs build`.

### CODE REVIEW

File: packages/core/src/facade.ts [hunk 1]
```diff
@@ -250,6 +250,216 @@ function isBufferUnchanged(
   );
 }
 
+type ParseAttempt<T> =
+  | {
+    readonly kind: "success";
+    readonly value: T;
+    readonly deferred?: true;
+    readonly deferredKeys?: DeferredMap;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Removed the unused success-branch fields in 4a74f840.

File: packages/core/src/facade.test.ts [hunk 2]
```diff
@@ -12059,6 +11996,149 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+          option: { names: ["--completions"] },
+          onShow: () => "COMP",
+        },
+        onError: () => "ERROR",
+        stdout: () => {},
+        stderr: () => {},
+      }),
+      { message: "--completions" },
+    );
+  });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts:12944]: Added the inline `--message=...` regression cases in 4a74f840.

File: packages/core/src/facade.ts [hunk 3]
```diff
@@ -1049,101 +1256,174 @@ function combineWithHelpVersion(
...
+
+  if (completion != null) {
     return {
-      type: "success",
-      value: "result" in parsedValue ? parsedValue.result : value,
+      type: "completion",
+      source: "option",
+      shell: completion.shell,
+      args: completion.args,
     };
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:1472]: Fixed in 4a74f840. Option-form completion now carries the matched command path through classification so subcommand completions stay scoped correctly.

File: packages/run/src/run.test.ts [hunk 4]
```diff
@@ -1558,6 +1558,18 @@ describe("runAsync", () => {
...
+      const parser = object({ value: argument(string()) });
+      const result = run(parser, {
+        args: ["help"],
+        programName: "myapp",
+        help: "command",
+        stdout: () => {},
+        stderr: () => {},
+      });
+      assert.deepEqual(result, { value: "help" });
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/run/src/run.test.ts]: Hardened both regression tests in 4a74f840 by providing explicit onExit handlers and asserting they were not called.

File: packages/core/src/facade.ts [hunk 5]
```diff
@@ -1049,101 +1256,174 @@ function combineWithHelpVersion(
...
+  let commandAction: MetaAction | undefined;
+  if (!hasConsumedPrefix && helpCommandNames.includes(firstArg)) {
+    commandAction = { index: 0, kind: "help" };
+  } else if (
+    !hasConsumedPrefix &&
+    versionCommandNames.includes(firstArg)
+  ) {
+    const boundaryIndex = completion?.index ?? optionArgs.length;
+    if (boundaryIndex <= 1) {
+      commandAction = { index: 0, kind: "version" };
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Fixed in 4a74f840. A leading `version` command now keeps precedence over a later completion option unless completion is the immediate next token.

File: CHANGES.md [hunk 6]
```diff
@@ -219,6 +219,14 @@ To be released.
...
     [[#231], [#782]]
 
+ -  Fixed `runParser()`, `runWith()`, and `runWithSync()` treating built-in
+    help, version, and completion tokens as globally reserved.  These runner
+    features are now parser-aware: command and option forms only trigger when
+    the user parser leaves them unconsumed, so positional values like `help`
+    and option values like `--help` can be parsed as ordinary data.  Startup
+    validation now also permits user parsers to reuse built-in meta names and
+    aliases.  [[#230], [#784]]
+
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md:229]: Updated the changelog entry in a49ee776 so it now matches the parser-aware collision behavior.

File: packages/core/src/facade.ts [hunk 7]
```diff
@@ -250,6 +250,216 @@ function isBufferUnchanged(
...
+      return {
+        kind: "failure",
+        error: message`Unexpected option or argument: ${context.buffer[0]}.`,
+        remainingArgs: context.buffer,
+        consumedCount: args.length - context.buffer.length,
+        optionsTerminated: context.optionsTerminated,
+        commandPath: getCommandPath(context.exec),
+      };
+    }
+  } while (context.buffer.length > 0);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:372]: I checked this against the current parser contract and I'm not taking this change. The top-level parser APIs intentionally continue parsing until the buffer is consumed or an error occurs, as documented in packages/core/src/parser.ts:847 and packages/core/src/parser.ts:943. Replacing the loop with a single parse pass here would diverge from `parseSync()` semantics and break multi-step parsers that legitimately consume input across successive parser states.

File: packages/core/src/facade.ts [hunk 8]
```diff
@@ -250,6 +250,216 @@ function isBufferUnchanged(
...
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:460]: I checked this against the current parser contract and I'm not taking this change. The top-level parser APIs intentionally continue parsing until the buffer is consumed or an error occurs, as documented in packages/core/src/parser.ts:943 and packages/core/src/parser.ts:1019. Replacing the loop with a single parse pass here would diverge from `parseAsync()` semantics and break multi-step parsers that legitimately consume input across successive parser states.

File: packages/core/src/facade.ts [hunk 9]
```diff
@@ -250,6 +250,216 @@ function isBufferUnchanged(
...
+    readonly trace?: ReturnType<typeof createInputTrace>;
+  },
+): ExecutionContext {
+  const runtime = createDependencyRuntimeContext();
+  return {
+    ...exec,
+    phase: "complete",
+    dependencyRuntime: runtime,
+    dependencyRegistry: runtime.registry,
+    commandPath: getCommandPath(context.exec) ?? exec.commandPath,
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:310]: I checked this one and I'm not changing it. At this call site the real bug was not the fallback expression in `createCompleteExec()`, but the option-form completion classification dropping `failure.commandPath`. That is fixed in 4a74f840. Changing this line alone would not preserve the matched subcommand for the failing path that triggered the review.

File: packages/core/src/facade.ts [hunk 10]
```diff
@@ -3130,56 +3220,63 @@ async function runWithBody<
...
+        programName,
+        args,
+        options,
+      ) as InferValue<TParser>,
     );
+  }
+  const augmentedParser1 = injectAnnotationsIntoParser(
+    parser,
+    phase1Annotations,
+  );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: I tried this refactor and reverted it. Reusing the same injected parser instance breaks the existing regression `runWithSync: two-phase fallback rebuilds the injected parser`, because the early-exit probe must not consume the same augmented parser object that the later parse pass relies on. The fresh injection here is intentional.

File: packages/core/src/facade.ts [hunk 11]
```diff
@@ -3382,42 +3479,43 @@ function runWithSyncBody<
...
-  // Two-phase parsing for two-pass contexts.
-  // First pass: parse with Phase 1 annotations
   const augmentedParser1 = injectAnnotationsIntoParser(
     parser,
     phase1Annotations,
   );
 
+  if (!needsTwoPhase) {
+    // All contexts are single-pass.
+    return runParser(augmentedParser1, programName, args, options);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: I tried this refactor and reverted it. Reusing the same injected parser instance breaks the existing regression `runWithSync: two-phase fallback rebuilds the injected parser`, because the early-exit probe must not consume the same augmented parser object that the later parse pass relies on. The fresh injection here is intentional.

**Comment by dahlia** [comment]
@codex review

File: packages/run/src/run.test.ts [hunk 12]
```diff
@@ -1558,6 +1558,24 @@ describe("runAsync", () => {
...
+        help: "command",
+        stdout: () => {},
+        stderr: () => {},
+        onExit: (code) => {
+          exited = true;
+          throw new Error(`UNEXPECTED_EXIT:${code}`);
+        },
+      });
+      assert.deepEqual(result, { value: "help" });
+      assert.equal(exited, false);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/run/src/run.test.ts]: Fixed in 30bb0e18. I replaced the remaining boolean-false equality assertions with the repo-preferred `assert.ok(!...)` form.

File: packages/core/src/facade.ts [hunk 13]
```diff
@@ -1049,101 +1251,183 @@ function combineWithHelpVersion(
...
+    };
+  }
 
-    // If version is present and help is not requested, show version
+  const optionArgs = failure.optionsTerminated ? [] : (() => {
+    const terminatorIndex = failure.remainingArgs.indexOf("--");
+    return terminatorIndex >= 0
+      ? failure.remainingArgs.slice(0, terminatorIndex)
+      : failure.remainingArgs;
+  })();
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:1407]: Fixed in d16d4766. `classifyOptionMeta()` now scans only the pre-`--` segment for option-form meta detection, but slices completion payload from the full remaining argv so tokens after `--` stay opaque completion input. I also added regressions for both separated and equals-form `--completion` payloads.

File: packages/core/src/facade.ts [hunk 14]
```diff
@@ -3130,56 +3228,63 @@ async function runWithBody<
...
+        programName,
+        args,
+        options,
+      ) as InferValue<TParser>,
     );
+  }
+  const augmentedParser1 = injectAnnotationsIntoParser(
+    parser,
+    phase1Annotations,
+  );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:3328]: I checked this against the current code and I'm not taking this refactor. Reusing the same injected parser instance here breaks the existing regression `runWithSync: two-phase fallback rebuilds the injected parser`, because the non-early-exit path intentionally rebuilds the phase-one wrapper after the early-exit probe. The duplicate injection is deliberate to keep the probe path and the real phase-one parse isolated.

File: packages/core/src/validate.test.ts [hunk 15]
```diff
@@ -526,246 +475,15 @@ describe("validateMetaNameCollisions", () => {
...
-        allOptions: e,
-        allCommands: e,
-        allLiterals: new Set(["--help=verbose"]),
-      },
-      [["option", "help option", ["--help"]]],
-    );
+    validateMetaNameCollisions([["command", "completion command", [
+      "--completion",
+    ]]]);
   });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/validate.test.ts]: Renamed in 30bb0e18. The test now describes the exact behavior it exercises instead of referring to `prefixMatch`.

File: packages/core/src/facade.ts [hunk 16]
```diff
@@ -250,6 +250,210 @@ function isBufferUnchanged(
...
+    const previousBuffer = context.buffer;
+    context = result.next;
+    if (isBufferUnchanged(previousBuffer, context.buffer)) {
+      return {
+        kind: "failure",
+        error: message`Unexpected option or argument: ${context.buffer[0]}.`,
+        remainingArgs: context.buffer,
+        consumedCount: args.length - context.buffer.length,
+        optionsTerminated: context.optionsTerminated,
+        commandPath: getCommandPath(context.exec),
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts:369]: Fixed in e3deb8a4. attemptParseSync() and attemptParseAsync() now preserve consumed progress in both failure paths, and I added sync and async regressions to keep consumed help tokens from being reclassified as runner meta.

File: packages/core/src/facade.test.ts [hunk 17]
```diff
@@ -3708,6 +3715,46 @@ describe("Subcommand help edge cases (Issue #26 comprehensive coverage)", () =>
...
+      runParser(parser, "myapp", ["--completion", "bash", "--", "b"], {
+        completion: {
+          option: true,
+        },
+        stdout: (text) => {
+          completionOutput += text;
+        },
+      });
+
+      assert.deepEqual(completionOutput.split("\n").sort(), ["beta", "build"]);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts]: Fixed in fe2ec74f. Those assertions now filter blank lines before sorting, so a trailing newline does not introduce a spurious empty completion entry.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.test.ts [hunk 18]
```diff
@@ -11395,6 +11645,160 @@ describe("branch coverage: facade.ts edge cases", () => {
...
+        usage: [],
+        leadingNames: new Set(["--help"]),
+        acceptingAnyToken: false,
+        initialState: undefined,
+        parse(context) {
+          return {
+            success: true as const,
+            next: context,
+            consumed: ["--help"],
+          };
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.test.ts:11891]: I checked this against the current code and I'm not taking this change. The second fixture is intentionally modeling the synthesized failure path in `attemptParse*()` where `parse()` reports consumed tokens but leaves the buffer unchanged, so the runner trips the `isBufferUnchanged(...)` guard and still needs to preserve `result.consumed.length` in the failure classification. Advancing `next.buffer` here would bypass that branch and only exercise the later `complete()` failure path, which is a different case from the one this regression is covering.

**Comment by dahlia** [comment]
Follow-up on https://github.com/dahlia/optique/pull/784#pullrequestreview-4082494615: fixed in 55a7c0b5. The phase-two seed extraction path now preserves commandPath while rebuilding the execution context, and I added sync/async regressions that verify two-pass contexts keep the matched subcommand scope.

**Comment by dahlia** [comment]
Follow-up on https://github.com/dahlia/optique/pull/784#pullrequestreview-4082537161: fixed in 9e0c6bf2. Option-form help reached after parser-consumed command tokens now stays on the user command doc path instead of switching to the built-in help, version, or completion command docs, and the new regression covers all three colliding command names.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/facade.ts [hunk 19]
```diff
@@ -3130,56 +3276,63 @@ async function runWithBody<
...
-  if (!needsTwoPhase) {
-    // All contexts are single-pass.
-    // Inject annotations into the parser's initial state
-    const augmentedParser = injectAnnotationsIntoParser(
-      parser,
-      phase1Annotations,
+  // Early exit: skip phase-two processing for genuine help/version/
+  // completion requests, but only after phase-1 annotations have been
+  // injected because they may change what the parser accepts as ordinary data.
+  if (await needsEarlyExitAsync(earlyExitParser, args, options)) {
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/facade.ts]: Addressed in f7d20a9d. `runWith()` and `runWithSync()` now skip the parse-only early-exit probe unless the run actually has two-phase contexts and configured help, version, or completion handlers, and I added sync and async regressions for the duplicate-parse paths.

File: packages/run/src/run.test.ts [hunk 20]
```diff
@@ -1558,6 +1558,24 @@ describe("runAsync", () => {
...
+        help: "command",
+        stdout: () => {},
+        stderr: () => {},
+        onExit: (code) => {
+          exited = true;
+          throw new Error(`UNEXPECTED_EXIT:${code}`);
+        },
+      });
+      assert.deepEqual(result, { value: "help" });
+      assert.ok(!exited);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/run/src/run.test.ts]: Addressed in 2bf7b4c2. The regression now calls runAsync() directly, so the test matches both the suite it lives in and the API it is meant to exercise.

**Comment by dahlia** [comment]
@codex review
